calico/felix/test/test_fetcd_api.py

```diff
@@ -0,0 +1,157 @@
+        """
+        self.etcd_api = EtcdAPI(self.m_config)
+        with patch.object(self.etcd_api.client, 'write'):
+            self.etcd_api.write_to_etcd('key', 'value', async=True)
+            self.run_actor_loop()
+            self.assertTrue(self.etcd_api.client.write.called)
+
+    def test_update_felix_status_disabled(self):
+        """
+        Test reporting is disabled for reporting interval 0
+        """
+        self.finish_setup()
+        gevent.sleep(1)
+        self.assertFalse(self.etcd_api.write_to_etcd.called)
+
+    def test_update_felix_status_single(self):
+        """
+        Test felix status is updated
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=15,
+                          REPORTING_TTL_SECS=37)
+
+        hostname = self.etcd_api._config.HOSTNAME
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        gevent.sleep(1)
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+        self.etcd_api.write_to_etcd.assert_has_calls([status_call,
+                                                      uptime_call])
+
+    def test_update_felix_status_continuous(self):
+        """
+        Test felix status is being continuously updated
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=3,
+                          REPORTING_TTL_SECS=10)
+        hostname = self.etcd_api._config.HOSTNAME
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        gevent.sleep(50)
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+        self.etcd_api.write_to_etcd.assert_has_calls(16 * [status_call,
+                                                           uptime_call])
+
+    @patch('calico.felix.fetcd._reconnect')
+    def test_update_felix_status_reconnects_on_etcd_exception(self, _reconnect):
+        """
+        Test felix status handles exceptions
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=4,
+                          REPORTING_TTL_SECS=12)
+        self.etcd_api.write_to_etcd = Mock(side_effect=EtcdException)
+
+        gevent.sleep(1)
+
+        self.assertTrue(_reconnect.called)
+
+
+class TestIfStatus(object):
+    """
+    Used to check whether status has expected format
+    i.e. whether timestamp is in ISO 8601 Zulu format
+    """
+    def __eq__(self, other):
+        import re
+        timestamp_regex = re.compile('.*"status_time": "\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z?".*')
+        is_timestamp =  timestamp_regex.match(str(other)) is not None
+        return is_timestamp
+    def __repr__(self):
+        return "<TestIfStatus()>"

```

Comment:
Two tips for \__repr__():

1. This should return a string which, if evaluated, returns an equal object to this one. So you shouldn't have the angle brackets; this should just return `TestIfStatus()`.
2. Rather than hard-coding the name of the class, you can access it through a special parameter: `self.__class__.__name__`. Using this makes the class more robust to subclassing, because the subclass can just inherit repr(). Unlikely to matter here, but it's a good habit to get into.

So I'd suggest writing the repr() (and the one below) like this:

    def __repr__(self):
        return '%s()' % self.__class__.__name__